core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooLock.java

```diff
@@ -123,189 +145,358 @@ public void unableToMonitorLockNode(Exception e) {
+    if (createdNodeName != null) {
+      String pathToDelete = path + "/" + createdNodeName;
+      LOG.debug("[{}] Failed to acquire lock in tryLock(), deleting all at path: {}", vmLockPrefix,
+          pathToDelete);
+      zooKeeper.recursiveDelete(pathToDelete, NodeMissingPolicy.SKIP);
+      createdNodeName = null;
     }
 
     return false;
   }
 
-  private synchronized void lockAsync(final String myLock, final AsyncLockWatcher lw)
-      throws KeeperException, InterruptedException {
+  /**
+   * Sort list of ephemeral nodes by their sequence number. Any ephemeral nodes that are not of the
+   * correct form will sort last.
+   *
+   * @param children
+   *          list of ephemeral nodes
+   * @return list of ephemeral nodes that have valid formats, sorted by sequence number
+   */
+  public static List<String> validateAndSortChildrenByLockPrefix(List<String> children) {
+    List<String> validChildren = new ArrayList<>();
+    children.forEach(c -> {
+      LOG.trace("Validating {}", c);
+      if (c.startsWith(ZLOCK_PREFIX)) {
+        String candidate = c.substring(ZLOCK_PREFIX.length() + 1);
+        if (candidate.contains("#")) {
+          int idx = candidate.indexOf('#');
+          String uuid = candidate.substring(0, idx - 1);
+          String sequenceNum = candidate.substring(idx + 1);
+          try {
+            LOG.trace("Testing uuid format of {}", uuid);
+            UUID.fromString(uuid);
+            if (sequenceNum.length() == 10) {
+              try {
+                LOG.trace("Testing number format of {}", sequenceNum);
+                Integer.parseInt(sequenceNum);
+                validChildren.add(c);
+              } catch (NumberFormatException e) {
+                LOG.warn("Child found with invalid sequence format: {} (no a number)", c);
+              }
+            } else {
+              LOG.warn("Child found with invalid sequence format: {} (not 10 characters)", c);
+            }
+          } catch (IllegalArgumentException e) {
+            LOG.warn("Child found with invalid UUID format: {}", c);
+          }
+        } else {
+          LOG.warn("Child found with invalid format: {} (does not contain second '#')", c);
+        }
+      } else {
+        LOG.warn("Child found with invalid format: {} (does not start with {})", c, ZLOCK_PREFIX);
+      }
+    });
 
-    if (asyncLock == null) {
-      throw new IllegalStateException("Called lockAsync() when asyncLock == null");
+    validChildren.sort(new Comparator<String>() {
+      @Override
+      public int compare(String o1, String o2) {
+
+        // Lock should be of the form:
+        // zlock#UUID#sequenceNumber
+        // Example:
+        // zlock#44755fbe-1c9e-40b3-8458-03abaf950d7e#0000000000
+        int secondHashIdx = 43;
+        return Integer.valueOf(o1.substring(secondHashIdx))
+            .compareTo(Integer.valueOf(o2.substring(secondHashIdx)));
+      }
+    });
```

Comment:
Could be shorter w/ a lambda.  Not sure if it compiles.  Could move the comment to the beginning of the method.

```suggestion
    validChildren.sort((o1,o2) -> {
        int secondHashIdx = 43;
        return Integer.valueOf(o1.substring(secondHashIdx))
            .compareTo(Integer.valueOf(o2.substring(secondHashIdx)));
    });
```